### PR TITLE
Add window_statistic utility

### PR DESCRIPTION
**Pull request by loroberge**
This PR includes a utility that calculates a local statistic on each node of a grid (mentioned in #1262 ). A user-input formula is calculated on a user-input field within a moving window, returning a result for each node on a grid. 

The common case of calculating local relief from a topographic elevation field is given as an example in the doc string.

### CODE REVIEW

**Comment by loroberge** [comment]
@nathanlyons I'm not sure where to request a reviewer, so tagging you here!

**Comment by nathanlyons** [comment]
@loroberge Looks great! I will work through a review over the next few days. I will let you know when I am done. Cannot wait to have this Landlab.

**Comment by gregtucker** [comment]
+1

**Review by nathanlyons — COMMENTED** [review]
@loroberge I provide comments both below and inline in the submitted files. 

- Please include the utility in __init__.py file located in landlab/utils. You can use add_halo as an example because this module has one function similar to your module. This enables your utility to be imported as a part of Landlab, i.e. `from landlab.utils import calculate_window_statistic`
- Add the utility to the docs. Doing so will add the utility to Landlab's readthedocs pages. Again you can follow add_halo.
    - Include the name of it here: landlab/docs/source/reference/utils/index.rst
    - Create a rst for it.
- Lastly, make sure your code follows established style guidelines. You can runs tools to do this automatically. See [this readthedocs page](https://landlab.readthedocs.io/en/master/development/practices/style_conventions.html).

Please let me know if I can clarify any of my comments.

File: landlab/utils/window_statistic.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    --------
+    >>> import numpy as np
+    >>> from landlab import RasterModelGrid
+    >>> grid = RasterModelGrid((5, 5), xy_spacing=10.0)
+    >>> grid.set_closed_boundaries_at_grid_edges(False,True,False,True)
+    >>> z = grid.add_zeros("topographic__elevation", at="node")
+    >>> z += np.arange(len(z))
+    
+    # Calculate relief using np.ptp function.
+    >>> relief = calculate_window_statistic(grid,'topographic__elevation',
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py:109]: I think I've solved this.

File: landlab/utils/window_statistic.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    **kwargs : optional
+        Keyword arguments that must be passed to func in addition to the field.
+    
+    Returns
+    -------
+    output : array
+        Output array containing the calculated values of the statistic.
+        Same length as input field. 
+    
+    Examples
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py:98]: done. I've added the new example and updated the previous examples to the same dimensions as described in the documentation.

File: landlab/utils/window_statistic.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    calculated at each iteration of the for loop.
+    A future version will act only on a subset of nodes.
+
+    Parameters
+    ----------
+    grid : ModelGrid
+        Landlab ModelGrid object.
+    field : Field attributed to ModelGrid nodes
+        User-input field on which to calculate the statistic of interest.
+        Must exist in grid.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py:76]: done

File: landlab/utils/window_statistic.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+import numpy as np
+
+from landlab import FieldError
+
+def calculate_window_statistic(grid,field,func,search_radius=100,
+                               calc_on_closed_nodes=True,**kwargs):
+    """Calculate a statistic using a function within a search window.
+    This works on any ModelGrid type.
+
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py:53]: great! this makes things very clear. done

**Review by loroberge — COMMENTED** [review]
Thanks for the detailed review! I've addressed all of your comments about the tool itself. I am still working on the unit tests for voronoi and radial grids.

File: landlab/utils/window_statistic.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+"""
+
+import numpy as np
+
+from landlab import FieldError
+
+def calculate_window_statistic(grid,field,func,search_radius=100,
+                               calc_on_closed_nodes=True,**kwargs):
+    """Calculate a statistic using a function within a search window.
+    This works on any ModelGrid type.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    nodes_in_loop = grid.nodes.flatten()
+    nodes_to_include = np.ones(grid.number_of_nodes,dtype=bool)
+    
+    if calc_on_closed_nodes is False:
+        closed_nodes = grid.status_at_node == grid.BC_NODE_IS_CLOSED
+        nodes_in_loop = nodes_in_loop[~closed_nodes]
+        nodes_to_include[closed_nodes] = False
+        output[closed_nodes] = np.NaN
+        
+    # Calculate "dist to point" then local relief at nodes within radius
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    Currently this utility is not optimized to run quickly on large grids
+    (e.g. 250x250 or larger), as the distance to every node on the grid is
+    calculated at each iteration of the for loop.
+    A future version will act only on a subset of nodes.
+
+    Parameters
+    ----------
+    grid : ModelGrid
+        Landlab ModelGrid object.
+    field : Field attributed to ModelGrid nodes
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+    Currently this utility is not optimized to run quickly on large grids
+    (e.g. 250x250 or larger), as the distance to every node on the grid is
+    calculated at each iteration of the for loop.
+    A future version will act only on a subset of nodes.
+
+    Parameters
+    ----------
+    grid : ModelGrid
+        Landlab ModelGrid object.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    field : Field attributed to ModelGrid nodes
+        User-input field on which to calculate the statistic of interest.
+        Must exist in grid.
+    func : function
+        User-input function that acts on the field.
+    search_radius : float, optional (defaults to 100 m)
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+"""
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py:2]: done

File: landlab/utils/window_statistic.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    Parameters
+    ----------
+    grid : ModelGrid
+        Landlab ModelGrid object.
+    field : Field attributed to ModelGrid nodes
+        User-input field on which to calculate the statistic of interest.
+        Must exist in grid.
+    func : function
+        User-input function that acts on the field.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    search_radius : float, optional (defaults to 100 m)
+        Radius of window within which the statistic is calculated.
+    calc_on_closed_nodes : boolean, optional (defaults to True)
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        Toggle calculation over all nodes including closed nodes (True) or all
+        nodes except closed nodes (False)
+    **kwargs : optional
+        Keyword arguments that must be passed to func in addition to the field.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    
+    Returns
+    -------
+    output : array
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

File: landlab/utils/window_statistic.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+from landlab import FieldError
+
+def calculate_window_statistic(grid,field,func,search_radius=100,
+                               calc_on_closed_nodes=True,**kwargs):
+    """Calculate a statistic using a function within a search window.
+    This works on any ModelGrid type.
+
+    Currently this utility is not optimized to run quickly on large grids
+    (e.g. 250x250 or larger), as the distance to every node on the grid is
+    calculated at each iteration of the for loop.
```
[the remark this answers is not in the document]
Comment (loroberge) [reply landlab/utils/window_statistic.py]: done

**Comment by nathanlyons** [comment]
@loroberge Almost there! Please let me know if you run into issues with the additional tests.

File: tests/utils/test_window_statistic.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+                                        calc_on_closed_nodes=False)
+
+    check = [     np.nan,  np.nan,  np.nan,  np.nan,
+               np.nan,     6.,     7.,     7.,    np.nan,
+            np.nan,    11.,    12.,    12.,    11.,  np.nan,
+               np.nan,     7.,     7.,     6.,    np.nan,
+                  np.nan,  np.nan,  np.nan,  np.nan]
+    
+    np.testing.assert_equal(relief, check)
+
```
[the remark this answers is not in the document]
Comment (loroberge) [reply tests/utils/test_window_statistic.py:30]: Done!

**Comment by loroberge** [comment]
> @loroberge Almost there! Please let me know if you run into issues with the additional tests.

@nathanlyons I've finished the unit tests.... Finally! Let me know if you have any other comments.

**Review by nathanlyons — DISMISSED** [review]
@loroberge The thorough tests look great. Thanks for the contribution!

**Comment by nathanlyons** [comment]
@mcflugen I get the following error when running make pretty from the main landlab direction :

```
error: cannot format /Users/njlyons/repos/landlab/landlab/landlab/__init__.py: module 'blib2to3.pygram' has no attribute 'python_grammar_no_print_statement_no_exec_statement_async_keywords'
```

This error prints multiple times except the files listed changes, I assume 444 times given this statement at the end of printing:

```
Oh no! 💥 💔 💥
62 files left unchanged, 464 files failed to reformat.
make: *** [pretty] Error 123
```

Any thoughts?

**Review by nathanlyons — APPROVED** [review]

**Comment by nathanlyons** [comment]
@mcflugen Ideas on why the notebooks and tests are failing? These notebooks and tests are not affiliated with this PR.

**Comment by loroberge** [comment]
@mcflugen I'm running into a docstring indentation error and a news fragment error. I think I've fixed both. Can you let me know what needs to be done next?

**Comment by mcflugen** [comment]
@loroberge I've fixed the indentation errors and the lint. Your news fragment is empty, though. Please add a description of your new feature, which will appear in the changelog. You can have a look at other news fragments in the news folder to see what it should look like.

**Review by mcflugen — APPROVED** [review]
👍 Looks good. Thanks @loroberge!

**Comment by loroberge** [comment]
> 👍 Looks good. Thanks @loroberge!

Great! Thank you, @mcflugen!!!

**Pull request merged**